skimage/feature/_texture.pyx

```diff
@@ -12,24 +12,38 @@ from .._shared.transform cimport integrate
 cdef extern from "numpy/npy_math.h":
     double NAN "NPY_NAN"
 
-
-def _glcm_loop(cnp.uint8_t[:, ::1] image, double[:] distances,
+ctypedef fused any_int:
+    cnp.uint8_t
+    cnp.uint16_t
+    cnp.uint32_t
+    cnp.uint64_t
+    cnp.int8_t
+    cnp.int16_t
+    cnp.int32_t
+    cnp.int64_t
+        
+def _glcm_loop(any_int[:, ::1] image, double[:] distances,
                double[:] angles, Py_ssize_t levels,
                cnp.uint32_t[:, :, :, ::1] out):
     """Perform co-occurrence matrix accumulation.
 
     Parameters
     ----------
     image : ndarray
-        Input image, which is converted to the uint8 data type.
+        Integer typed input image. Only positive valued images are supported. 
+        If type is other than uint8, the argument `levels` needs to be set.
     distances : ndarray
         List of pixel pair distance offsets.
     angles : ndarray
         List of pixel pair angles in radians.
     levels : int
-        The input image should contain integers in [0, levels-1],
+        The input image should contain integers in [0, `levels`-1],
```

Comment:
It is possible to calculate a GLCM for negative valued images, but the indices of the resulting matrix would not correspond to the grey levels anymore. Then we would need to give back the map indices vs. grey levels (which would change the API). Or we always give back the full data range matrix, starting with the negative values. In practice however, a full data range matrix for types > uint8 does not fit in the RAM of most computers, so this is not a good solution neither. 

If we want to do this properly, we could do binning prior to GLCM calculation (see discussion with soupault). But would this still be inline with the current API? I have the impression that this would change the function quite dramatically. Binning makes a lot of sense for this function, but it can also be left to the user.